client/starwhale/api/_impl/dataset/model.py

```diff
@@ -250,7 +250,17 @@ def __repr__(self) -> str:
         return f"Dataset: {self._uri.name}, loading uri: {self._loading_uri}, pending commit uri: {self._pending_commit_uri}"
 
     def __len__(self) -> int:
-        return self._total_rows
+        """__len__ slow but accurate"""
+        if self._dataset_builder is None:
+            return self.approximate_size
+        else:
+            self.flush()
+            return self._dataset_builder.calculate_rows_cnt()
```

Comment:
Is it better to add a cache for the non-dirty dataset?

Reply:
updated